File: app/src/main/java/social/entourage/android/authentication/login/slideshow/LoginSlideshowFragment.java
package social.entourage.android.authentication.login.slideshow;


import android.content.res.Resources;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.content.res.AppCompatResources;
import androidx.fragment.app.Fragment;
import androidx.viewpager.widget.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import social.entourage.android.EntourageEvents;
import social.entourage.android.R;
import social.entourage.android.authentification.login.slideshow.LoginSlideshowPageAdapter;


/**
 * Login slideshow {@link Fragment} subclass.
 */
public class LoginSlideshowFragment extends Fragment {

    // ----------------------------------
    // CONSTANTS
    // ----------------------------------

    public static final String TAG = LoginSlideshowFragment.class.getSimpleName();

    // ----------------------------------
    // ATTRIBUTES
    // ----------------------------------

    @BindView(R.id.login_slideshow_view)
    ViewPager pagerView;

    @Nullable
    @BindView(R.id.login_slideshow_indicator_layout)
    LinearLayout indicatorLayout;

    private List<ImageView> dots;

    // ----------------------------------
    // LIFECYCLE
    // ----------------------------------

    public LoginSlideshowFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View v = inflater.inflate(R.layout.login_slideshow, container, false);
        ButterKnife.bind(this, v);

        return v;
    }

    @Override
    public void onViewCreated(@NonNull final View view, @Nullable final Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initializeCarousel();
    }

    private void initializeCarousel() {
        LoginSlideshowPageAdapter pageAdapter = new LoginSlideshowPageAdapter(getChildFragmentManager());
        pagerView.setAdapter(pageAdapter);
        pagerView.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(final int position, final float positionOffset, final int positionOffsetPixels) {

            }

            @Override
            public void onPageSelected(final int position) {
                EntourageEvents.logEvent(EntourageEvents.EVENT_LOGIN_SLIDESHOW);
                selectDot(position);
            }

            @Override
            public void onPageScrollStateChanged(final int state) {

            }
        });

        addDots();
    }

    private void addDots() {
        if (indicatorLayout == null) return;
        dots = new ArrayList<>();

        ImageView dot1 = indicatorLayout.findViewById(R.id.slideshow_b1);
        dots.add(dot1);

        ImageView dot2 = indicatorLayout.findViewById(R.id.slideshow_b2);
        dots.add(dot2);

        ImageView dot3 = indicatorLayout.findViewById(R.id.slideshow_b3);
        dots.add(dot3);

        ImageView dot4 = indicatorLayout.findViewById(R.id.slideshow_b4);
        if (dot4 != null) dots.add(dot4);
    }

    private void selectDot(int idx) {
        if (dots == null || dots.size() <= idx) return;
        for(int i = 0; i < LoginSlideshowPageAdapter.NUM_PAGES; i++) {
            int drawableId = (i==idx)?(R.drawable.carousel_bullet_filled):(R.drawable.carousel_bullet_empty);
            Drawable drawable = AppCompatResources.getDrawable(getContext(), drawableId);
            dots.get(i).setImageDrawable(drawable);
        }
    }

}

File: app/src/main/java/social/entourage/android/authentication/login/slideshow/LoginSlideshowFragment.kt
package social.entourage.android.authentication.login.slideshow

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.appcompat.content.res.AppCompatResources
import androidx.fragment.app.Fragment
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import kotlinx.android.synthetic.main.login_slideshow.*
import social.entourage.android.EntourageEvents
import social.entourage.android.R
import java.util.*

/**
 * Login slideshow [Fragment] subclass.
 */
class LoginSlideshowFragment  : Fragment() {
    // ----------------------------------
    // ATTRIBUTES
    // ----------------------------------
    private val dots: ArrayList<ImageView> = ArrayList()
    // ----------------------------------
    // LIFECYCLE
    // ----------------------------------
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.login_slideshow, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initializeCarousel()
    }

    private fun initializeCarousel() {
        login_slideshow_view?.let { pager ->
            pager.adapter = LoginSlideshowPageAdapter(childFragmentManager)
            pager.addOnPageChangeListener(object : OnPageChangeListener {
                override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}
                override fun onPageSelected(position: Int) {
                    EntourageEvents.logEvent(EntourageEvents.EVENT_LOGIN_SLIDESHOW)
                    selectDot(position)
                }

                override fun onPageScrollStateChanged(state: Int) {}
            })
        }
        addDots()
    }

    private fun addDots() {
        login_slideshow_indicator_layout?.let { layout ->
            dots.clear()
            layout.findViewById<ImageView>(R.id.slideshow_b1)?.let { dots.add(it) }
            layout.findViewById<ImageView>(R.id.slideshow_b2)?.let { dots.add(it) }
            layout.findViewById<ImageView>(R.id.slideshow_b3)?.let { dots.add(it) }
            layout.findViewById<ImageView>(R.id.slideshow_b4)?.let { dots.add(it) }
        }
    }

    private fun selectDot(idx: Int) {
        if (dots.size <= idx) return
        for (i in 0 until LoginSlideshowPageAdapter.NUM_PAGES) {
            val drawableId = if (i == idx) R.drawable.carousel_bullet_filled else R.drawable.carousel_bullet_empty
            val drawable = AppCompatResources.getDrawable(requireContext(), drawableId)
            dots[i].setImageDrawable(drawable)
        }
    }

    companion object {
        // ----------------------------------
        // CONSTANTS
        // ----------------------------------
        val TAG = LoginSlideshowFragment::class.java.simpleName
    }
}